### PR TITLE
[DAG] Fold freeze(shuffle(x,y,m)) -> shuffle(freeze(x),freeze(y),m)

### PR DESCRIPTION
**Pull request by RKSimon**
If the shuffle mask contains no undef elements, then we can move the freeze through a shuffle node.

This requires special case handling to create a new ShuffleVectorSDNode.

### CODE REVIEW

**Comment by uweigand** [comment]
SystemZ change LGTM.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -15482,6 +15499,24 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
     }
   }
 
+  // Special case handling for ShuffleVectorSDNode nodes.
```
Comment (bjope) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:15502]: Why isn't it enough to add VECTOR_SHUFFLE to AllowMultipleMaybePoisonOperands?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:15502]: I'll take another look - this patch started out a little different and I think we can reduce the diff even further.

**Comment by RKSimon** [comment]
I've been able to simplify this a lot more than I expected - including adding isGuaranteedNotToBeUndefOrPoison handling

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -15533,8 +15534,16 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
     if (Op.getOpcode() == ISD::UNDEF)
       Op = DAG.getFreeze(Op);
   }
-  // NOTE: this strips poison generating flags.
-  SDValue R = DAG.getNode(N0.getOpcode(), SDLoc(N0), N0->getVTList(), Ops);
+
+  SDValue R;
+  if (auto *SVN = dyn_cast<ShuffleVectorSDNode>(N0)) {
```
Comment (bjope) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:15539]: Wouldn't the else part work just fine also for VECTOR_SHUFFLE?
The mask is in Ops[2].

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:15539]: ShuffleVectorSDNode embeds the mask in the node itself - its not an operand

Comment (bjope) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:15539]: Ok. Got it.

**Review by bjope — APPROVED** [review]
LGTM

**Pull request merged**